include/index_helper.cuh

```diff
@@ -1106,4 +1106,302 @@ namespace quda {
     return (((x[3]*X[2] + x[2])*X[1] + x[1])*X[0] + x[0]) >> 1;
   }
 
+
+  /**
+   * These are index helper functions used in the order classes of openQCD, i.e.
+   *
+   * - OpenQCDOrder       in quda:include/gauge_field_order.h
+   * - OpenQCDDiracOrder  in quda:include/color_spinor_field_order.h
+   * - OpenQCDOrder       in quda:include/clover_field_order.h.
+   *
+   * The main helper functions are ipt() and iup(), giving pure function
+   * implementations of the ipt[] and iup[][] arrays (see
+   * openqcd:include/global.h) that are needed to calculate the correct offsets
+   * of the fields base pointers.
+   */
+  namespace openqcd {
+
+    /**
+     * @brief      Returns the surface in direction mu
+     *
+     * @param      X     Extent in all 4 directions
+     * @param[in]  mu    Direction
+     *
+     * @return     Surface
+     */
+    __device__ __host__ inline int surface(const int *X, const int mu)
+    {
+      if (mu==0) {
+        return X[1]*X[2]*X[3];
+      } else if (mu==1) {
+        return X[0]*X[2]*X[3];
+      } else if (mu==2) {
+        return X[0]*X[1]*X[3];
+      }
+      return X[0]*X[1]*X[2];
+    }
+
+
+    /**
+     * @brief      Return BNDRY (see openqcd:include/global.h)
+     *
+     * @param[in]  L      Local lattice extent L0-L3 in txyz convention
+     * @param[in]  nproc  NPROC0-NPROC3 from openqcd
+     *
+     * @return     BNDRY
+     */
+    __device__ __host__ inline int bndry(const int *L, const int *nproc)
+    {
+      return 2*(((1-(nproc[0]%2))*surface(L, 0))
+              + ((1-(nproc[1]%2))*surface(L, 1))
+              + ((1-(nproc[2]%2))*surface(L, 2))
+              + ((1-(nproc[3]%2))*surface(L, 3))
+      );
+    }
+
+
+    __device__ __host__ inline int ifc(const int *L, const int *nproc, const int mu)
```

Comment:
Could you add a small comment what this does?

Reply:
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